lib/engine/game/g_18_royal_gorge/step/buy_train.rb

```diff
@@ -16,6 +16,21 @@ def room?(entity)
             # even when train tight, there's room for a self-rust
             return true if entity.trains.any? { |t| t.rusts_on == upcoming_train.name }
           end
+
+          def check_connected!(buyer, train)
+            return if train.owner == @game.depot
+
+            seller = train.corporation
+            seller_nodes = @game.graph.connected_nodes(seller)
+            return if @game.graph.connected_nodes(buyer).any? { |node, _| seller_nodes.include?(node) }
```

Comment:
Would it be more efficient to do `return unless (@game.graph.connected_nodes(train.corporation) & @game.graph.connected_nodes(buyer)).empty?`

Reply:
`graph.connected_nodes` is a hash, and so doesn't have a `&` method. You could do it as:
```
@game.graph.connected_nodes(train.corporation).keys.intersect?(@game.graph.connected_nodes(buyer).keys)
```
But I'm not sure if that's clearer.

Reply:
Since the return of `connected_nodes` is a hash and we're looking at the keys, `seller_nodes.include?(node)` inside the loop is O(1) while the outer loop is O(buyer's).

When I first read these comments I was thinking the same as @crericha, that what I originally wrote was O(buyer's * seller's), which it would be if they were arrays instead of hashes. I didn't realize that with hashes it was fine until I just pushed my latest version, which is what @ollybh suggested, only split up across multiple lines.

Hopefully this version does the job--it is linear time, and hopefully that is more clear when reading it.